rust/network/src/reliable_sender.rs

```diff
@@ -0,0 +1,247 @@
+    /// It returns a vector of cancel handlers with no specific order.
+    pub async fn lucky_broadcast(
+        &mut self,
+        mut addresses: Vec<SocketAddr>,
+        data: Bytes,
+        nodes: usize,
+    ) -> Vec<CancelHandler> {
+        addresses.shuffle(&mut self.rng);
+        addresses.truncate(nodes);
+        self.broadcast(addresses, data).await
+    }
+}
+
+/// Simple message used by `ReliableSender` to communicate with its connections.
+#[derive(Debug)]
+struct InnerMessage {
+    /// The data to transmit.
+    data: Bytes,
+    /// The cancel handler allowing the caller task to cancel the transmission of this message
+    /// and to be notified of its successfully transmission.
+    cancel_handler: oneshot::Sender<Bytes>,
+}
+
+/// A connection is responsible to reliably establish (and keep alive) a connection with a single peer.
+struct Connection {
+    /// The destination address.
+    address: SocketAddr,
+    /// Channel from which the connection receives its commands.
+    receiver: Receiver<InnerMessage>,
+    /// The initial delay to wait before re-attempting a connection (in ms).
+    retry_delay: u64,
+    /// Buffer keeping all messages that need to be re-transmitted.
+    buffer: VecDeque<(Bytes, oneshot::Sender<Bytes>)>,
+}
+
+impl Connection {
+    fn spawn(address: SocketAddr, receiver: Receiver<InnerMessage>) {
+        tokio::spawn(async move {
+            Self {
+                address,
+                receiver,
+                retry_delay: 200,
+                buffer: VecDeque::new(),
+            }
+            .run()
+            .await;
+        });
+    }
+
+    /// Main loop trying to connect to the peer and transmit messages.
+    async fn run(&mut self) {
+        let mut delay = self.retry_delay;
+        let mut retry = 0;
+        loop {
+            match TcpStream::connect(self.address).await {
+                Ok(stream) => {
+                    info!("Outgoing connection established with {}", self.address);
+
+                    // Reset the delay.
+                    delay = self.retry_delay;
+                    retry = 0;
+
+                    // Try to transmit all messages in the buffer and keep transmitting incoming messages.
+                    // The following function only returns if there is an error.
+                    let error = self.keep_alive(stream).await;
+                    warn!("{}", error);
+                }
+                Err(e) => {
+                    warn!("{}", NetworkError::FailedToConnect(self.address, retry, e));
+                    let timer = sleep(Duration::from_millis(delay));
+                    tokio::pin!(timer);
+
+                    'waiter: loop {
+                        tokio::select! {
+                            // Wait an increasing delay before attempting to reconnect.
+                            () = &mut timer => {
+                                delay = min(2*delay, 60_000);
```

Comment:
do we reset delay when we succeed or the connection is now slower?

Reply:
yes we reset the timer upon successful connection